Update hashbrown

thaliaarchi opened the pull request:
[`hashbrown` 0.15](https://github.com/rust-lang/hashbrown/blob/master/CHANGELOG.md#v0150---2024-10-01) was released this month, which notably removed the `RawTable` API in favor of `HashTable` and changed its hasher from `ahash` to the faster [`foldhash`](https://github.com/orlp/foldhash).

- Migrate to `HashTable`
- Update `indexmap` to bump its `hashbrown` dependency
- Refactor a manual implementation of an index map with cloning in `TermDag` to `indexmap::IndexMap`

Besides this, Max's [`symbol_table`](https://github.com/mwillsey/symbol_table) still requires an older version of `hashbrown` in the lockfile. I have a draft to update that and could rebase to include it, but didn't figure out how to run its Criterion benchmarks.

saulshanabrook reviewed (dismissed):
Thanks for this PR!

Overall this looks good, I like the simplification of the TermDag from this change.

In terms of the benchmarks, we just added them and are still refining their use. I don't think the slowdowns listed are significant. I have an open PR (https://github.com/egraphs-good/egglog/pull/444) to turn off these shorter benchmarks, which have high variability due to indeterminism in the memory allocator.

If you click on the [details](https://codspeed.io/egraphs-good/egglog/branches/thaliaarchi:update-hashbrown) for one of the "slowdowns" you can see that it's due to allocation during parsing:

![Screenshot 2024-10-18 at 9 55 13 AM](https://github.com/user-attachments/assets/16713bc8-20ee-4369-972d-68dccf76e907)

The longer running benchmarks, like eggcc-extraction and math-microbenchmark seem to have no change from this PR.

Alex-Fischman commented:
+1 to not letting "performance regressions" in the parser block this PR

thaliaarchi commented on src/termdag.rs:
```diff
@@ -21,14 +21,8 @@ pub enum Term {
-    // - every key of hashcons is in nodes
-    pub nodes: Vec<Term>,
-    pub hashcons: HashMap<Term, TermId>,
+    /// A bidirectional map between deduplicated `Term`s and indices.
+    nodes: IndexSet<Term>,
```
Would it be better to create getters for what you need? Then it matters less what it's stored as. Since there were no other uses of the field, it looked safe to make private; a method would also make that more clear. `pub` is fine too, though 🤷‍♀️

saulshanabrook replied:
Yeah, that sounds good to me.

Looking through the usages in Python, it seems like the only things I do with a termdag is [call `term_to_expr`](https://github.com/egraphs-good/egglog-python/blob/12d8d5f35f35a578a909a7605a40695d25bdd8ab/src/lib.rs#L15C13-L15C25) and then [lookup a term based on its id (`termdag.nodes[term_id]`)](https://github.com/egraphs-good/egglog-python/blob/12d8d5f35f35a578a909a7605a40695d25bdd8ab/python/egglog/egraph_state.py#L562)

thaliaarchi replied:
Looking up a term by `TermId` is already exposed via `TermDag::get`. So it sounds like what you need is already public?

What about the conversion in `convert_struct!` in your earlier comment? Would you need a `fn TermDag::iter(&self) -> impl Iterator<Item = &Term>`?

saulshanabrook replied:
Naw that's ok, I can just not expose the TermDag struct like that, and instead just expose a few methods on it.

thaliaarchi commented:
I've pushed a commit, which replaces symbol_table with my fork that bumps their version of hashbrown. It's a hack to run benchmarks in CI. I'll then submit a PR to symbol_table, depending on the results, and rebase here.

yihozhang commented:
cykjson is a small cool egglog example that does the [CYK parsing algorithm](https://en.wikipedia.org/wiki/CYK_algorithm) of JSON-like strings. It is a more Datalog-like workload (dynamic programming) with some e-class manipulations

thaliaarchi commented:
I dropped the commit for benchmarking updating `symbol_table`. If there's a response there later, it can be updated here separately.

I also changed `fn TermDag::get(&self, id: TermId) -> Term` to return `&Term` instead. All internal usages use the result by reference, so cloning is unnecessary. Does this affect anything externally?

Alex-Fischman commented:
> I also changed fn TermDag::get(&self, id: TermId) -> Term to return &Term instead. All internal usages use the result by reference, so cloning is unnecessary. Does this affect anything externally?

This is good, egglog is unstable and users can clone if they need it.

saulshanabrook commented:
Thanks @thaliaarchi for working on this and responding to all the feedback! If you have anything to add, we are also discussing the tradeoffs with hash performance and determinism in this post: https://github.com/egraphs-good/egglog/pull/439#issuecomment-2435724938

thaliaarchi commented:
@saulshanabrook Thanks! I'm glad to see such speedups!
